### PR TITLE
Multinodes examples

### PR DESCRIPTION
**Pull request by camillemarini**
Reworking examples to consider multiple nodes (by default 2 Nodes) to "fix" #25

### CODE REVIEW

File: examples/titanic/scripts/add_train_algo_random_forest.py [hunk 1]
```diff
@@ -49,38 +56,39 @@
...
+    data_keys = client.get_dataset(data_manager_key)['trainDataSampleKeys']
+    traintuple = client.add_traintuple({
+        'algo_key': algo_key,
+        'objective_key': objective_key,
+        'data_manager_key': data_manager_key,
+        'train_data_sample_keys': data_keys,
+        'in_models_keys': [traintuple_key]
+    }, exist_ok=True)
+    traintuple_key = traintuple.get('key') or traintuple.get('pkhash')
+    print(traintuple_key)
```
Comment (inalgnu) [inline examples/titanic/scripts/add_train_algo_random_forest.py]: was it a debug print here or is it wanted?

Comment (camillemarini) [reply examples/titanic/scripts/add_train_algo_random_forest.py]: thanks @inalgnu it was indeed a debug print :-)

**Comment by camillemarini** [comment]
Hello @jmorel thank you for your feedback. I agree with you that it would be better to have done it in separate PR :-) 
However, given the time constraints, I won't be available to create two separate PR for this one, and it would be very helpful if it could be reviewed as it. 
Cheers

File: examples/cross_val/scripts/cross_val_algo_random_forest.py [hunk 2]
```diff
@@ -60,7 +92,7 @@
     # testtuple
     testtuple = client.add_testtuple({
         'traintuple_key': traintuple_key,
-        'data_manager_key': dataset_key,
+        'data_manager_key': data_manager_key,
```
Comment (jmorel) [inline examples/cross_val/scripts/cross_val_algo_random_forest.py]: Due to the recent changes in the backend (optional evaluation), the objective key should be moved from the definition of the traintuple to the testtuple's.

File: examples/cross_val/scripts/cross_val_algo_random_forest.py [hunk 3]
```diff
@@ -16,31 +16,63 @@
...
+        {
+            'train_data_sample_keys': list(X[train_index]),
+            'test_data_sample_keys': list(X[test_index])
+        } for train_index, test_index in kf.split(X)
+    ]
+}
+
+# saving folds keys
+with open(os.path.join(current_directory, folds_keys_path), 'w') as f:
+    json.dump(folds_keys, f, indent=2)
```
Comment (jmorel) [inline examples/cross_val/scripts/cross_val_algo_random_forest.py]: The scripts fails with the following error when run from the cross_val directory:
```
Traceback (most recent call last):
  File "scripts/cross_val_algo_random_forest.py", line 67, in <module>
    with open(os.path.join(current_directory, folds_keys_path), 'w') as f:
FileNotFoundError: [Errno 2] No such file or directory: 'scripts/scripts/../folds_keys.json'
```
The file name should probably just be `folds_keys_path`.

File: examples/titanic/scripts/add_train_algo_random_forest.py [hunk 4]
```diff
@@ -49,38 +56,39 @@
...
-    'data_manager_key': assets_keys['dataset_key'],
-    'train_data_sample_keys': assets_keys['train_data_sample_keys']
-}, exist_ok=True)
-traintuple_key = traintuple.get('key') or traintuple.get('pkhash')
-assert traintuple_key, 'Missing traintuple key'
+for data_manager_key in data_manager_keys:
+    data_keys = client.get_dataset(data_manager_key)['trainDataSampleKeys']
+    traintuple = client.add_traintuple({
+        'algo_key': algo_key,
+        'objective_key': objective_key,
```
Comment (jmorel) [inline examples/titanic/scripts/add_train_algo_random_forest.py]: The objective key should be moved from the traintuple to the testtuple.

File: examples/compute_plan/scripts/add_compute_plan.py [hunk 5]
```diff
@@ -56,10 +70,11 @@
...
 testtuples = []
 previous_id = None
-for train_data_sample_key in train_data_sample_keys:
+# for data_btrain_data_sample_key in train_data_sample_keys:
+for data_batch in data_batches:
     traintuple = {
-        'data_manager_key': dataset_key,
-        'train_data_sample_keys': [train_data_sample_key],
+        'data_manager_key': data_batch[0],
+        'train_data_sample_keys': data_batch[1],
```
Comment (jmorel) [inline examples/compute_plan/scripts/add_compute_plan.py]: With the new structure of compute plans, objectives must be included directly into testtuples and algos directly into traintuples.

**Comment by samlesu** [comment]
@camillemarini Some work has been done by @jmorel to handle a multi nodes example in this [PR](https://github.com/SubstraFoundation/substra/pull/154 ). Would that be ok for you to close this PR and merge Jérémy's work?

**Comment by camillemarini** [comment]
@samlesu That's fine for me!

**Pull request closed without merging**